packages/dds/tree/src/feature-libraries/chunked-forest/basicChunk.ts

```diff
@@ -0,0 +1,369 @@
+
+    private getOffsetPath(offset: number): UpPath | undefined {
+        const length = this.indexStack.length - offset;
+        if (length === 0) {
+            return undefined; // At root
+        }
+
+        assert(length > 0, "invalid offset to above root");
+        assert(length % 2 === 0, "offset path must point to node not field");
+
+        // Perf Note:
+        // This is O(depth) in tree.
+        // If many different anchors are created, this could be optimized to amortize the costs.
+        // For example, the cursor could cache UpPaths from the anchorSet when creating an anchor,
+        // then reuse them as a starting point when making another.
+        // Could cache this at one depth, and remember the depth.
+        // When navigating up, adjust cached anchor if present.
+
+        let path: UpPath | undefined;
+        // Skip top level, since root node in path is "undefined" and does not have a parent or index.
+        for (let height = 2; height < length; height += 2) {
+            const key = this.getStackedFieldKey(height - 1);
+            path = {
+                parent: path,
+                parentIndex: this.getStackedNodeIndex(height),
+                parentField: key,
+            };
+        }
+
+        path = {
+            parent: path,
+            parentIndex: offset === 0 ? this.index : this.getStackedNodeIndex(length),
+            parentField: this.getStackedFieldKey(length - 1),
+        };
+        return path;
+    }
+
+    public enterField(key: FieldKey): void {
+        assert(this.mode === CursorLocationType.Nodes, "must be in nodes mode");
+        this.siblingStack.push(this.siblings);
+        this.indexStack.push(this.index);
+
+        // For fields, siblings are only used for key lookup and
+        // nextField and which has arbitrary iteration order,
+        // so making a array of just key here works.
+        // This adds an allocation, so it's optimizing code simplicity and for the other use case (enumeration)
+        // at the cost of an allocation here.
+        this.index = 0;
+        this.siblings = [key];
+    }
+
+    public nextField(): boolean {
+        this.index += 1;
+        if (this.index === (this.siblings as []).length) {
+            this.exitField();
+            return false;
+        }
+        return true;
+    }
+
+    public firstField(): boolean {
+        const fields = this.getNode().fields;
+        if (fields.size === 0) {
+            return false;
+        }
+
+        this.siblingStack.push(this.siblings);
+        this.indexStack.push(this.index);
+        this.index = 0;
+        this.siblings = [...fields.keys()]; // TODO: avoid this copy
+        return true;
+    }
+
+    public seekNodes(offset: number): boolean {
+        assert(this.mode === CursorLocationType.Nodes, "can only seekNodes when in Nodes");
+        assert(this.indexOfChunk < this.siblings.length, "out of bounds indexOfChunk");
+
+        this.indexWithinChunk += offset;
```

Comment:
Since this is specifically for BasicChunks, do we ever expect any more than one node within the chunk? It seems like topLevelLength should always be 1 so this seems like we should either generalize this as a ChunkCursor or just use StackCursor for BasicChunkCursor.

Reply:
To be able to efficiently support long sequences (with non O(n) editing and shallow copying) we will eventually need some support for grouping contents in fields into chunks hierarchically. For the current feature set though, you are right that its unneeded. I have added a TODO comment up in the constructor where this is declared about this.